File: advent-of-code/2020/day19/day19a.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Advent of Code 2020, day 19, part A

$VERBOSE = 1

require 'English'
require 'optparse'

# Extending Enumerator with a take_until method, analogous to take_while.
class Enumerator
  def take_until
    take_while { |item| !yield(item) }
  end
end

def each_line_in_stanza
  if block_given?
    each_line_in_stanza.each { |line| yield line }
  else
    ARGF.each_line.lazy.map(&:strip).take_until(&:empty?)
  end
end

show_pattern = false
show_matches = false

OptionParser.new do |parser|
  parser.on '-p', '--pattern', 'Show the constructed pattern' do
    show_pattern = true
  end
  parser.on '-g', '--grep', 'Show the matching lines (like grep)' do
    show_matches = true
  end
  parser.on '-h', '--help', 'Show options help' do
    puts parser
    exit
  end
end.parse!

rules = {}

make_nonterminal_rule = lambda do |id, template|
  lambda do
    rules[id] = -> { raise "cyclic dependency for rule #{id}" }

    alternatives = template.map do |row|
      row.map { |k| rules[k].call }.join
    end

    pattern =
      if alternatives.size == 1
        alternatives.first
      else
        "(?:#{alternatives.join('|')})"
      end

    rules[id] = -> { pattern }
    pattern
  end
end

add_rule = lambda do |id, expr|
  if expr =~ /^"([^"]+)"$/
    literal, = $LAST_MATCH_INFO.captures
    escaped = Regexp.escape(literal)
    rules[id] = -> { escaped }
  else
    branches = expr.split(/\s+\|\s+/) # alternation

    template = branches.map do |branch|
      branch.split(/\s+/).map(&:to_i) # concatentation
    end

    rules[id] = make_nonterminal_rule.call(id, template)
  end
end

each_line_in_stanza do |rule|
  id_digits, expr = rule.split(/:\s+/)
  add_rule.call(id_digits.to_i, expr)
end

pattern = rules[0].call

if show_pattern
  puts pattern
  puts
end

line_regex = /^#{pattern}$/

count = each_line_in_stanza.count do |text|
  is_match = text.match?(line_regex)
  puts text if show_matches && is_match
  is_match
end

puts if show_matches
puts count

File: advent-of-code/2020/day19/day19a.pl
#!/usr/bin/env perl
# Advent of Code 2020, day 19, part A

use strict;
use warnings;
use 5.026;
use feature qw(say);
use experimental qw(signatures);
use Carp qw(croak);
use Getopt::Long qw(:config bundling);

sub each_line_in_stanza :prototype(&) ($code) {
    while (<<>>) {
        s/\s+$//msx;
        s/^\s+//msx;
        last if length == 0;
        $code->();
    }
    return;
}

my $show_pattern = 0;
my $show_matches = 0;

GetOptions('pattern|p' => \$show_pattern, 'grep|g' => \$show_matches)
    or exit 1;

my %rules;

my sub make_nonterminal_rule($id, @template) {
    return sub {
        $rules{$id} = sub { croak "cyclic dependency for rule $id" };

        my @alternatives = map {
            join q{}, map { $rules{$_}->() } @{$_}
        } @template;

        my $unparenthesized = join q{|}, @alternatives;

        my $pattern = (scalar @alternatives == 1 ? $unparenthesized
                                                 : "(?:$unparenthesized)");

        $rules{$id} = sub { $pattern };
        return $pattern;
    };
}

my sub add_rule($id, $expr) {
    if (my ($literal) = $expr =~ /^"([^"]+)"$/msx) {
        my $escaped = quotemeta $literal;
        $rules{$id} = sub { $escaped };
    } else {
        # Split alternation into a list of branches.
        my @branches = split /\s+[|]\s+/msx, $expr;

        # Map each branch to the row of ids of subpatterns for concatenation.
        my @template = map { [map { $_ + 0 } split /\s+/msx] } @branches;

        $rules{$id} = make_nonterminal_rule($id, @template);
    }
    return;
}

each_line_in_stanza {
    my ($id_digits, $expr) = split /:\s+/msx;
    add_rule($id_digits + 0, $expr);
};

my $pattern = $rules{0}->();
if ($show_pattern) { print "$pattern\n\n" }
my $line_regex = qr/^$pattern$/msx;

my $count = 0;
each_line_in_stanza {
    if ($_ =~ $line_regex) {
        ++$count;
        if ($show_matches) { say };
    }
};

if ($show_matches) { print "\n" }
say $count;
